Update specification

fulldecent opened the pull request:
This PR creates implementable specifications for Plus Codes and Open Location Codes with minimal changes to what is currently published.

And it removes all contradictory information in this project.

In the process, this solves a majority of open issues in one pull request.

Because of contradictions in the project as it, it is NOT possible to make this into bite-sized PRs. I appreciate your time reviewing this thoroughly.

## Normative changes

* Define codes for north and south poles
* Specify which code levels of full codes may be shortened to which code levels of short codes
* Specify which API methods are REQUIRED or OPTIONAL
* Establish and define naming (Plus Codes, area, bounds, long code, etc.) for consistent usage
* Define which full code a short code refers to
* Move any authoritative normative statements from wiki and other places here into the specifications
  * wiki/Naming-guidelines
  * FAQ.txt
  * olc_definition.adoc
  * specification.md
  * CHANGELOG.md
  * Guidance-for-shortening-codes.md
  * README.md / "Each implementation provides..." notes

## Other chages

* Removed duplicative/outdated specifications and "recommendations", updated links
* Moved all documentation into Documentation folder, updated links
* Basic copyediting and formatting on files I touched
* Migrated documentation consistently to Markdown format
* Updated all references to use "Plus Codes" when codes are being referred to (implementations and systems are still "Open Location Code"), fixes [Plus Codes capitalization inconsistent with marketing website](https://github.com/google/open-location-code/issues/441)
* Move implementations into Implementations folder, delete non-implementations, reference everything in README with note of new API compliance
* Added test cases to match normative requirements
* Adds notes about which platforms will produce wrong results because they have insufficient number precision (useful for embedded applications)

## Fixed issues

- ["Global code" may refer to a 10 or 11 digit code](https://github.com/google/open-location-code/issues/438)
  - Because references to the outdated "global code" and regex are gone
- [OLC inconsistent reference to Earth equatorial circumference](https://github.com/google/open-location-code/issues/449)
  - Because equatorial distance is no longer referenced, and more approximate distances are given
- [Specify a unique representation for North and South poles](https://github.com/google/open-location-code/issues/450)
  - Because this is specified in the Area and bounds section
- [Specification should recommend function names](https://github.com/google/open-location-code/issues/432)
  - Because adds recommended method names
- [Documentation conflicts on whether 8 digits may be removed in shortening](https://github.com/google/open-location-code/issues/462)
  - Because this is specified as 2, 4, or 6 characters are removed
- [Plus Codes regex is incorrect](https://github.com/google/open-location-code/issues/438)
  - Because new correct regexes are given
- [OLC inconsistent reference to Earth equatorial circumference](https://github.com/google/open-location-code/issues/449)
  - Because no circumference is referenced and values in meters are approximated slightly
- [OLC specification doesn't strictly state validity status of codes with odd length](https://github.com/google/open-location-code/issues/196)
  - Because specification is given on code lengths
- [Short codes with padding are unintentionally considered valid](https://github.com/google/open-location-code/issues/273)
  - Because specification is given for short codes
- [Google maps and 'Find your code' tool create short codes which aren't decodable offline](https://github.com/google/open-location-code/issues/343)
  - Beacuse invalid marketing claims are removed
- [Test for efficient implementation by fulldecent](https://github.com/google/open-location-code/pull/444)
  - Because tests are added and implementations are required to implement correctly.

## Makes progress on these issues

* [Plus Codes capitalization inconsistent with marketing website](https://github.com/google/open-location-code/issues/441)
  * Because it uses and specifies the correct capitalization of "Plus Codes". But the implementations also need to be updated.
* [Floating point representation errors may cause incorrect codes](https://github.com/google/open-location-code/issues/307)
  * Because detailed discussion includes notes about floating point errors and it includes a recommendation implementation that produces correct results. It also references classes of floating point errors that are not discussed in this issue. But the implementations also need to be updated.
* [Use consistent naming in this repository -- plus codes or open location code](https://github.com/google/open-location-code/issues/431)
  * Because the documentation is updated to make consistent and correct usage of Plus Code versus Open Location Code. Implementations are not updated. I hope this PR can please be reviewed and merged before I consider the implementations.

## :star: Requested action

- [ ] Merge this pull request

- [ ] Update project wiki per commit https://github.com/fulldecent/open-location-code-wiki/pull/1/files. Project maintainer can do this with:
  1. Clone git@github.com:google/open-location-code.wiki.git
  2. Pull from https://github.com/fulldecent/open-location-code-wiki
  3. `git log` and/or `git diff HEAD~` to review changes
  4. Push

- [ ] Tag and release this update at https://github.com/google/open-location-code/releases

## Follow-on work

After this PR is merged, I can:

- [ ] Create reference implementation, validate and fix one existing implementation and note is as verified compliant
- [ ] Move implementations to Implementations/ folder (also update README, .travis.yml)
- [ ] Delete non-implementations (just delete file, already properly referenced in README)
- [ ] Move demos to Demos/ folder
- [ ] Losslessly compress images

And I can help but might need help with:

- [ ] Create example implementation for embedded applications in 32 bits and 16 bits

fulldecent commented:
This PR includes a move of all implementations into the Implementations/ folder and this inflates to a scary high Files Changed number. Please review this PR using VS code and you will see there are fewer actual changes. This file moving is required for this PR because it distinguishes the "implementations" from the "demos". And implementations now have a full specification.

I have worked hard to make this a "bite sized" and "minimal" PR.

fulldecent commented:
Requesting PR review, please, from @bilst, @bocops, @drinckes

bocops commented:
I haven't had time to check all of this in detail. The fact that this is a massive change that also moves around unrelated parts makes it hard to do so - and considering that even much smaller PR currently aren't accepted quickly makes me wonder if this repo and its contributors are even willing and able to handle a PR of this scope at the moment.

For what it's worth, reading through the rewritten `Plus Codes Specification.md`, the whole thing feels much less readable and understandable to me. While more strict specifications are appreciated in some places, what this now lacks is a good overview of what a plus code even is, what it looks like, what it can be used for, without immediately jumping into "SHALL NOT", "MUST", and "MAY" legalese.

On another note, the rewritten specs claim to be v1.0.5, a patch-level bump from the previous 1.0.4, but I immediately see some areas where there are larger (if not even breaking) changes to the specification. This means that this new specification, if accepted, should be 1.1.0 or even 2.0.0, which in turn means that there would need to be an even more thorough look at it (and reason for the change in the first place).

To give just one example for a potentially breaking change, areas around the poles are now strictly defined to include or not include the poles themselves. In the case of the south pole, where before all areas having the pole as one of their corner points would include it, now only one of them (for any given code length) does. On top of that, the area that does is the one east of the antimeridian, not the one east of the prime meridian, although the latter ("0°") is typically used when giving coordinates of the poles. While this might look like a rather unimportant detail, it can make the resulting code worse than strictly necessary, as has been laid out in #450.

fulldecent commented:
👍 Thank you for looking at my PR, I do really appreciate it.

# Poles

I agree that all details here are important and I commit to fix everything to earn acceptance of this PR. Yes, prime meridian is a more customary choice.

✅ Fixed at 8345cf4234d26ffedcfa3201e541875ed8181bf8

# Introduction and readability

The [current Plus Codes specification](https://github.com/google/open-location-code/blob/main/docs/specification.md) also does not include an introduction.

And this PR does include [the same introduction file](https://github.com/google/open-location-code/blob/618ad5bff4603dd2c4f540a696a5e7599c489c29/Documentation/Introduction.md) as before.

Is this the file you were looking for?

Here is a potential brief addition at the top of the Plus Codes specification (above the version history):

> In much of the world, street addresses are poorly defined or non-existent. Plus Codes allow to encode location information into easily exchangeable codes. For example, `8FVC2200+` includes the area around Romoos, Switzerland. Shorter and longer codes can represent various area sizes, are helpful for navigation and are very convenient for humans to remember and use.

❓  Shall I make that addition?

# Massive changes

The PR currently moves all implementations into an Implementations/ folder. This is in contrast to Demos which go to the Demos/ folder. This is significant because the Open Location API Specification applies to "implementations".

❓ Shall I factor out this change to discuss separately?

# Version number

You are referring to [Semantic Versioning](https://semver.org). Thank you, I agree 1.1.0, at a minimum, is appropriate.

🦶 Updated at d7a7a91540e3da9911717a21e980bc57bfe9580b.

❓ Shall I change this 2.0.0?

# Corners and edges

I have corrected the specification to include the south-west corner of each area. I tested the C implementation works this way and also added test cases.

✅ Corrected in 2ce7dca3a8281774aa46c4bb0034bd8a0b94db00

# Other notes

I am not deterred that other PRs are not accepted quickly.

This project ["strongly encourages technical contributions"](https://github.com/google/open-location-code/blob/main/CONTRIBUTING.md#contributing-to-open-location-code) and I hope this significant technical PR, pushed by somebody committed to see it through, and flexible to get it done, can see the light of day. I am also around to take on more responsibility later when I know good PRs are actively reviewed and acted on. (Specifically, I'd like to update one of the implementations to be a "reference implementation" as a great starting example.)

bocops commented:
Re: Introduction/Readability ("Shall I make that addition?"): Perhaps I've just been confused by the amount of change. Let's revisit this at a later point, once all the other pieces have fallen into place, wherever that might be?

Re: Massive changes ("Shall I factor out this change to discuss separately?"): I actually appreciate the enhanced structure of having all implementations but not the random rest in one folder, but I think this might best be handled separately, if only to remove some of the noise from this PR.

Re: Version number: You're right, this repo doesn't explicitly state that it _does_ use semantic versioning, but I think that is the reasonable choice. Whether a minor version bump would be sufficient in my opinion depends (probably among other things) on whether decoding a plus code encoded with 1.0.4 using these new specifications still leads to the same coordinates in all instances. Perhaps let's identify those changes first that would mandate a major version bump, and see which ones are considered absolutely necessary?

fulldecent commented:
# Massive changes

All file moves were extracted from this PR to discuss another day. (Separate discussion referenced in Follow-On Work on top post.)

✅ Fixed in aef4682c5e50b4ddf0f78d0a49facd8474d7d8b4

# Version number

I believe the only normative note is that north and south poles are defined in PR to have a specific Plus Code at each level.

The current spec canon (i.e. wiki, "spec", and other documents in this repo that make statements) refer to the poles as:
- Being unimportant
- Having multiple representations for the South Pole
- Having zero and also many representations for the North Pole

Every other input and output is intended to be the same in this PR. And I have tested the C implementation for conformance for hand-checked inputs and it passed. (Commit at 2ce7dca3a8281774aa46c4bb0034bd8a0b94db00.)

Depending on your definition of "change" this PR represents either a "change" for one location and a new specification for the other; or it represents a "don't care" for one location and a "resolving conflict" for the other.

Next is a more philosophical note, maybe. The proposed spec uses math to define Plus Codes and warns about math bit precision. And the current spec canon uses a pseudocode as a spec. The current spec canon will produce different results on 64-, 32- and 16-bit (think embedded) systems. If this causes the PR to produce different results (on 32-, 16-bit systems) than the current spec canon, I consider it a feature not a bug.

bocops commented:
> I believe the only normative note is that north and south poles are defined in PR to have a specific Plus Code at each level.
> [...]
> Depending on your definition of "change" this PR represents either a "change" for one location and a new specification for the other; or it represents a "don't care" for one location and a "resolving conflict" for the other.

As far as the poles are concerned, I admit that this is a somewhat obscure detail - but as this specification rewrite is all about being precise and strict, we shouldn't be too handwavy about the details. As has already been stated in #450, currently someone might want to encode "the south pole as well as nearby locations along the anti-meridian", and end up with a plus code of 22222222+. Decoding this plus code under 1.0.4, the result will be an area that includes the south pole. Unter this new specification, it explicitly does not.

Now, I'm still not sure whether having a perfect tiling in the mathematical sense is important enough to go through all this - but if we do, we should be clear about the fact that at least one plus code decodes to different things under 1.0.4 specs and new specs.

Of course, there might also be other details outside just encoding/decoding that lead to certain version bumps, like the actually required API methods, or what regex is being suggested as matching valid codes.

bocops commented on Documentation/Plus Codes Specification.md:
```diff
@@ -0,0 +1,119 @@
+
+## Copyright
+
+The authors dedicate this document to the public domain under [CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/).
```
I'm not sure this is the proper way to copyright this, as it includes content that has previously been made available under Apache-2.0 license.

fulldecent replied:
Thank you! Deleted.

:white_check_mark: Fixed at d7c6d2507ce49c2a0ef0d7c1d5249ef4b668db37

bocops commented on Documentation/Plus Codes Specification.md:
```diff
@@ -0,0 +1,119 @@
+
+A Plus Code uses significant digits to encode its southern and western bounds. The significant digits of a Plus Code represent latitude (north) and then longitude (east) offsets from the starting location 90° S, 180° W. 
+
+A significant digit in digit places 1, 3, 5, 7 and 9 represents an offset north equal to the north-south distance for that code length (see the next higher code length in the code length table above) multiplied by the significant digit value.
```
Why "next higher" code length? If I have code length 4, I multiply the values in digit places 3 and 4 by 1°, not by (the "next higher" value of) 20°. If this is supposed to convey looking up (3+1)=4 in the table, I suggest rephrasing this to be more clear.

fulldecent replied:
Fixed at a2b41682dda221e01d6d42507c7860bcdb8f0e38

bocops commented on Documentation/Plus Codes Specification.md:
```diff
@@ -0,0 +1,119 @@
+
+The **center** of a Plus Code area is defined as the arithmetic mean of its bound's parallels and meridians.
+
+## Precision
```
Comment - this is not a problem with this PR, but still:

1. Is "Precision" even the correct term in this context, or should it be "Accuracy"?
2. Is it actually correct to give precision/accuracy as the "maximum side length", or should this rather be something like the diagonal (or even half of that, assuming that we're talking about distance to center?).

fulldecent replied:
Agreed, fixed at 33044f51f05234c96da1600ed24db9b5f46100a7

bocops commented on Documentation/Plus Codes Specification.md:
```diff
@@ -0,0 +1,119 @@
+4. (If the southern bound is 90° S and the western bound is 0° W) the South Pole (i.e. the point at 90° S); and
+5. (If the southern bound is not 90° S) on the southern bound from (including) the western bound to (excluding) the eastern bound.
+
+This specification references latitudes and longitudes on Earth under [WGS 84](https://earth-info.nga.mil), which is the standard used by the Global Positioning System. If another geodetic is used, that context MUST be agreed between the producer and consumer of a Plus Code.
```
Is it really the case / important that coordinates are generally referenced according to WGS84, or is it rather the case that producer and consumer need to agree on that? Suggestion: remove this here, but merge into a section on "necessary producer/consumer agreements" that also includes details about how shortening codes only works if producder and consumer can agree on reference locations.

fulldecent replied:
I think it is well understood that Plus Codes refer to places on Earth, offset based on the high noon sun location of Greenwich, England. Or in plain English, "these are GPS coordinates, and compatible with all your normal GPS stuff". Otherwise we are just talking about coordinates on a sphere. Happy to remove if that is already obvious to everyone, (but then should update Longest edge [formerly Precision] to reference Earth).

Currently the relevant sentence for short code Plus Codes is the first sentence in that section. 

> A **short code** Plus Code is meaningful only when the producer and consumer of the Plus Code can agree on the approximate latitude and longitude of a **reference location**.

Shall I add a note at the bottom of that section which details under which circumstances the producer and consumer will disagree an a shortening?

bocops replied:
I've actually already seen one application that uses plus codes for either Mars or the Moon. In the end, all that OLC does is to translate a 2-tuple of floating point numbers into a string and back again. The exact semantics of that process - whether it is current GPS coordinates on Earth, or coordinates in any other geodetic system on Earth, or coordinates on any other celestial body whether real or fictional - doesn't really matter outside of the necessary producer/consumer agreement.

The suggestion to specify "longest edge" to refer to Earth sounds good.

fulldecent replied:
Updated Earth note at 79ca2c2c94a42f7a44edeb0e52fc5e9524aac846

---

A shortening will work if the producer and consumer agree, to an approximate amount, of where a reference location is. Beyond some specific tolerance, the shortening/recovery will not be inverses.

You and I both understand this fuzziness but it is not spelled out. Shall I spell it out?

bocops commented on Documentation/Open Location Code API Specification.md:
```diff
@@ -0,0 +1,78 @@
+* Latitude inputs MUST treat a latitude greater than 90° N as if it were the North Pole (i.e. at 90° N).
+* Latitude inputs MUST treat a latitude lower than 90° S as if it were the South Pole (i.e. at 90° S).
+* Longitude inputs MUST treat values outside the range from (including) 180° W to (excluding) 180° E as if they were the equivalent longitude in this range.
+* Longitude inputs MUST NOT cause runtime performance linearly dependent on the longitude. E.g. use `LONGITUDE % 360`, not `WHILE (LONGITUDE > 180) LONGITUDE -= 360`.
```
Already added as a PR to yours, but now in review form: `LONGITUDE % 360` should be replaced by `MOD(LONGITUDE,360)` to avoid faulty implementations in languages where `%` is not the module but the remainder operator (Java being one such case).

fulldecent replied:
Fixed at 89063bda8c46c2a4964cc8d3b5b56c0e2a5b557e

bocops commented on Documentation/Open Location Code API Specification.md:
```diff
@@ -0,0 +1,78 @@
+Implementations MUST follow these REQUIREMENTS for all public methods:
+
+* Plus Code Inputs MUST treat non-upper-case inputs as if they were upper case.
+* Plus Code outputs MUST be uppercase and conform to the [Plus Codes Specification].
```
Comment: "MUST be uppercase" is more strict than "should return upper case characters" in current API.txt. (Similarly for inputs that MUST vs. should be accepted in non-major-case form). I agree that this new requirement is sensible, but it is an incompatible API change in terms of semver.

fulldecent replied:
It is currently specified that inputs are case insensitive at https://github.com/google/open-location-code/blob/a323e80159d6d6170158cc22497250cb86bf3583/docs/specification.md#library-implementation-requirements. Not a breaking change

Currently the spec references the digits for [Plus Codes] as upper case at https://github.com/google/open-location-code/blob/a323e80159d6d6170158cc22497250cb86bf3583/docs/specification.md#digits and the API specs say that a [Plus Code] is returned. One interpretation of these words is that outputs MUST be uppercase.

Good notes here, it helped me find and improve an inconsistency with the PCREs at 64ffd9c

bocops replied:
However, API.txt states that (emphasis mine):

> Note that any method that returns an Open Location Code **should** return
> upper case characters.
> 
> Methods that accept Open Location Codes as parameters **should** be case
> insensitive.

If different parts of the documentation disagree on something, that's bad. However, while that is the case I believe we can't really state that strictly mandating something that was optional at least to some readings of the previous documentation is **not** a breaking change.

fulldecent replied:
I can get behind that argument. Fixed at 2e8ac0abc41e71b30747a62970e7c177193ea41f

bocops commented on Documentation/Open Location Code API Specification.md:
```diff
@@ -0,0 +1,78 @@
+| `encode`           | Latitude, longitude, code length                             | The full code Plus Code with specified code length with area that contains the specified coordinates |
+| `decode`           | Full code Plus Code string                                   | The southern parallel and western meridian for the Plus Code area |
+| `shorten`          | Full code Plus Code, reference location latitude and longitude | The short code Plus Code representing the input where all significant digits that are allowed to be omitted are omitted; or if no digits are allowed to be omitted then the input is returned |
+| `recoverNearest`   | Short code Plus Code, reference location latitude and longitude | The unique full code Plus Code nearest (latitudinal distance + longitudinal distance) to the reference location |
```
In the output column, "(latitudinal distance + longitudinal distance)" seems to suggest using rectilinear distance instead of euclidean distance as the metric to determine the nearest location. Suggestion: mention the actual metric by name, or remove quoted text if an exact metric does not need to be mandated.

bocops commented on Documentation/Open Location Code API Specification.md:
```diff
@@ -0,0 +1,78 @@
+
+Note: it is possible to implement `isFull` using a Perl Compatible Regular Expression, see the Plus Code specification.
+
+## OPTIONAL public methods
```
Current API.txt does not seem to consider exactly the above three methods as REQUIRED and all remaining ones as OPTIONAL. I see the point of not forcing implementations to offer "shorten"-related functionality, but I think that at least `isValid` and perhaps `encode/decode` with variable code length might as well be required. Perhaps check what the current implementations do, then let's discuss?

fulldecent replied:
Also I moved `isValid`, which regards full+short codes from required to optional. This is consistent with all other things relating to short codes being optional. Separately, no implementation implements `isValid` currently as specified (see below).

Breaking change warning added.

🦶 : Progress at ce6bb6db03524857f7b3762b5c6d3236b02428b3 // marking below as an open item to review implementations and which methods should be required 

Replaced specifications:

* https://github.com/google/open-location-code/blob/main/API.txt
* https://github.com/google/open-location-code/blob/a323e80159d6d6170158cc22497250cb86bf3583/docs/specification.md#library-implementation-requirements

bocops replied:
Whether or not implementations  currently implement a required method correctly shouldn't play a role in determining whether or not that method should stay required.

Also, I think that making `isValid` optional without any kind of common replacement is kind of backwards. A valid code (full or short) is a valid code, whether or not any given implementation is able to handle it.

One hypothetical replacement might be `canBeDecoded`, a boolean that would return true iff the code is valid and full, OR if it is valid, short and the implementation in question implements the optional bits. However, this definition already uses a code's validity again, so the `isValid` method might as well keep being required.

fulldecent replied:
Fixed `isValid` at 362c8ecd975cebd037233f4f6411eea94818e134

bocops commented on Documentation/Open Location Code API Specification.md:
```diff
@@ -0,0 +1,78 @@
+
+| Method name          | Input(s)                        | Output(s)                                                    |
+| -------------------- | ------------------------------- | ------------------------------------------------------------ |
+| `isFull`             | Plus Code string                | True if this is a long code Plus Code, false otherwise       |
```
Already added as a PR to yours, but now in review form: "long code Plus Code" is not a defined term and should be replaced with "full code Plus Code".

Aside: This is a mouthful - can we perhaps define just "full code" and "short code" early on and from then on omit always adding "Plus Code"?

fulldecent replied:
I agreed // I had been overzealous in fully qualifying all references to "short code" and "full code". Fixed at a896494bc7fd0b640c71d04b0ac4bb7c1263f7a5

fulldecent commented on Documentation/Plus Codes Specification.md:
```diff
@@ -0,0 +1,119 @@
+* If both distances are less than 0.5 degrees, digit places 1–2 or 1–4 MAY be omitted.
+* If both distances are less than 1/40 degrees, digit places 1–2, 1–4 or 1–6 MAY be omitted.
+
+A short code represents the unique full code Plus Code nearest (latitudinal distance + longitudinal distance) to the reference location.
```
Yes. But I don't want people to think this means Manhattan distance in meters.

fulldecent commented on Documentation/Plus Codes Specification.md:
```diff
@@ -0,0 +1,119 @@
+
+The latitudinal distance (in degrees) and longitudinal distance (in degrees) between the Plus Code area's center and the reference location determines which digit places can be omitted in the short code representation:
+
+* If both distances are less than 10 degrees, digit places 1–2 MAY be omitted.
```
Yes! This is built into the spec under "Short codes" where it specifies the "if both distances..." part. These use half the distance as the precision amount for that code level.

I believe this amounts to 50% margin. Is that correct?

bocops replied:
It's true that, with an **exact** reference location agreed upon between both parties, a short code can be recovered within one full code area for any given code level (e.g. within +/- 10° if the full code was shortened by two digits). However, often the reference locations aren't exact, so to make short codes more useful many parts of the current documentation warn about pushing the limits where this isn't necessary.

For example, this wiki page mentions a safety margin of 20% (which I falsely remembered as a factor of 0.4 - should of course be 0.8): https://github.com/google/open-location-code/wiki/Guidance-for-shortening-codes

That page, by the way, is referenced from this repo's README, which also states:

> If the reference location is derived from a town or city name, it is dependent on the accuracy of the geocoding service. Although one service may place "Zurich" close to the Google office, another may move it by a hundred meters or more, and this could be enough to prevent the original code being recovered. Rather than a large city size feature to generate the reference location, it is better to use smaller, neighbourhood features, that will not have as much variation in their geocode results.
> 
> Guidelines for shortening codes are in the wiki.
> 
> Recovering shortened codes works by providing the short code and a reference location. This does not need to be the same as the location used to shorten the code, but it does need to be nearby.

fulldecent commented:
Thank you for all the notes. Below is a summary of (recently) resolved and (every) remaining items.

# Poles and breaking changes

Breaking change warning added.

:white_check_mark: Fixed at ffe0a3a4f4da2a15aa4b2fef72e7855d23a547be

# Required API methods

Also I moved `isValid`, which regards full+short codes from required to optional. This is consistent with all other things relating to short codes being optional. Separately, no implementation implements `isValid` currently as specified (see below).

Breaking change warning added.

:white_check_mark: Fixed at ce6bb6db03524857f7b3762b5c6d3236b02428b3

Replaced specifications:

* https://github.com/google/open-location-code/blob/main/API.txt
* https://github.com/google/open-location-code/blob/a323e80159d6d6170158cc22497250cb86bf3583/docs/specification.md#library-implementation-requirements

# Regex

Agreed, there is a slight difference in the `isFull`, `isShort`, `isValid` methods. Now they actually must validate the Plus Codes. This is easy because a PCRE implements this exactly.

No implementations implement the current required API. Specifically, they all check for the format separator (+) which is not part of the spec.

The old regexes were wrong and/or conflicting and removed.

The new Regexes exactly match valid Plus Codes, the spec, and the plain English meaning of the method name "is valid".

Breaking change warning added.

:white_check_mark: Fixed at daa1c353daef1818f10e797e22766e62cb21fd9c

# Copyright

Thank you! Deleted.

:white_check_mark: Fixed at d7c6d2507ce49c2a0ef0d7c1d5249ef4b668db37

# Code length lookup

:white_check_mark: Fixed at a2b4168

# "Precision"

A more specific wording describes the geometry of the Plus Code area rather than the unclear word "precision".

:white_check_mark: Fixed at 33044f51f05234c96da1600ed24db9b5f46100a7

# Open items:

* Where exactly is Greenwich? https://github.com/google/open-location-code/pull/463#discussion_r647330859
* Which methods should be required? https://github.com/google/open-location-code/pull/463#discussion_r647331111 // and should supporting all code lengths (not just length 10) be required?
* Short codes
  * Add test cases
  * Add a margin of safety
    * https://github.com/google/open-location-code/pull/463#discussion_r650514831
    * https://github.com/google/open-location-code/pull/463#discussion_r647330711
  * If geocoding, recomend to use city names // https://github.com/google/open-location-code/pull/463#discussion_r650514831

fulldecent commented:
@bocops Could you please share any thoughts on the open items immediately above?

bocops commented:
Sorry for the late reply. :)

**Regarding required vs. optional methods**, I think we agree that all current functionality regarding the handling of **full** codes needs to stay required. According to current API.txt, that is: `isFull`, `encode`, `decode`

I'm not sure that making all functionality related to **short** codes completely optional is the correct thing to do. At the very least, contributors should be strongly encouraged to offer this functionality in their implementations. Whether we decide for or against this change, this would affect these methods: `isShort`, `shorten`, `recoverNearest`

The remaining `isValid` is a special case. Even if we decide to make short code handling optional, a short code is still a valid code, so if this method name stays, it should behave the same across all implementations, independent of whether a certain implementation does or doesn't handle short codes. At that point, discussing the removal of shorten-related functionality becomes somewhat moot.

In the end, because of this, and because the main problem with short codes seems to be not their existence but the fact that the main publisher of short codes chooses reference locations rather opaquely, I believe that all seven methods might as well stay required.

**Regarding the question of required code length support**, I assume this means full code lengths outside of shortening? A while ago, we had a lengthy discussion about what code lengths are sensible in the first place, and the decision was made to restrict code length to <=15, because anything beyond that doesn't really have any real world use cases.

An argument could be made that codes referencing an area of <= 10cm² (code length 14, 15) aren't that sensible in practice, either - but unless those fall into the range of "technically impossible on specific systems", I don't think we should make some lengths required but others optional. If there are precision problems, we should rather define how an implementation should deal with them - for example, should decoding a "too long" code simply fail, or should it return the next-bigger area?

**Regarding safety margins when shortening**, I think the best description so far can be found on the wiki page [Guidance for shortening codes](https://github.com/google/open-location-code/wiki/Guidance-for-shortening-codes), although I'd prefer to not simply hand out magic numbers like 0.4, but explain a bit better why we're not going for the theoretical full range of 0.5.

fulldecent commented:
I am considering to update the PR and make it:

- **All methods required.** This is simpler and it restricts which implementations we link to/develop in this repo to only be the highest quality implementations. If somebody wants to make a lesser implementation, it will be up to THEM to explain why they chose to exclude short codes or something else.
- **Safety.** Clearly specify safety margins.
-**Required code length.** Like you say, only specify how it should work and how it should gracefully degrade. Explain why.

If I implement all these things, does that clear the threshold for merging this PR?

fulldecent commented:
@bocops @drinckes Can you please confirm that if I make the changes above https://github.com/google/open-location-code/pull/463#issuecomment-945004628 and create a reference implementation that this PR may be merged?

bocops commented:
@fulldecent This is a decision for @drinckes or potentially @bilst to make. I'm still not convinced that all of the breaking changes are really necessary, but in the end this is their project. Either way, I agree it would be nice to see some closure here, this has gone on long enough.

fulldecent commented:
@drinckes @bilst, please advise.

I'd like to update [these three things](https://github.com/google/open-location-code/pull/463#issuecomment-945004628) if you think it will be mergable. 

Then merged I will work on a reference implementation.

bilst commented:
Sorry for the slow reply! I think you point out some interesting theoretically valid technical issues, but given that they haven't been an actual problem for any users, I'm not sure it's worth the added complexity/formality to call them out specifically.

Easiest PRs to accept are small scope improvements to clarity (e.g. editing a paragraph or two in a single file). Any increased complexity of documentation or added requirements need to be weighed carefully against their potential to discourage adoption by GIS folks who might not be as comfortable working with more formal specifications.

fulldecent commented:
@bilst

> issues... haven't been an actual problem for any users

Instead of considering the problems of _users_ (which is super limited, because few people use Plus Codes), please consider the problems of _potential users_.

There are many people that don't use Plus Codes. Perhaps this is because it has problems and it is not possible to correctly implement Plus Codes.

People that develop products (even GIS folks) like correct specifications and example code. On the other hand, people that start developing products, but never finish... those are the people that like vague and inconsistent specifications.

This PR addresses these issues, holds Plus Codes in the highest light, and lays the path forward for multiple compatible and correct implementations.

---

I will not break this PR up into smaller pieces because too many things are broken as-is.

You are welcome to close this PR without merging if you still do not have the time to provide a proper review.

drinckes commented:
I'm not even going to try to understand a single PR of this size.

For future reference, I would rather have multiple (i.e. small) PRs for a single issue, than one PR that attempts to address multiple issues.
